### PR TITLE
Make secret management opt-in

### PR DESCRIPTION
**Pull request by acidmoose**
- Make use of java ServiceLoader to look for secret handlers
- S3 secret will now be dynamically loaded into Flags if the dependency to
is secrets-s3 is added to a project. If it is not added, it is not used.
- Parse secrets before parsing arguments
- Open secret fetching method if someone wants to use it to only fetch secrets
- Change secret naming scheme to secret:s3:bucket:key. This is more
in line with s3 concepts. Better than the fake s3 url previously used and is
much easier to parse.

### CODE REVIEW

File: flags/pom.xml [hunk 1]
```diff
@@ -1,68 +1,38 @@
...
-        <scope>import</scope>
-      </dependency>
+        <dependency>
+            <groupId>net.sf.jopt-simple</groupId>
+            <artifactId>jopt-simple</artifactId>
+            <version>5.0.3</version>
+        </dependency>
+        <dependency>
+            <groupId>com.bettercloud</groupId>
+            <artifactId>vault-java-driver</artifactId>
```
Comment (agavrilov76) [inline flags/pom.xml:26]: the vault driver is still in the dependency list for the main module?

File: flags/src/main/java/org/secondbase/flags/Flags.java [hunk 2]
```diff
@@ -363,13 +356,26 @@ private void addInstancedOption(
...
+     * Pass args through secret handling.
+     * @param args - Arguments passed from main method.
+     * @return this
+     */
+    public String[] fetchSecrets(final String[] args) {
+        String[] ret = args;
+        for (final SecretHandler secretHandler : secretHandlers) {
+            ret = secretHandler.fetch(args);
+        }
+        return ret;
```
Comment (agavrilov76) [inline flags/src/main/java/org/secondbase/flags/Flags.java]: Will it work if args param has secrets from multiple security providers?

File: flags/src/main/java/org/secondbase/secrets/SecretHandlerException.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package org.secondbase.secrets;
+
+import java.io.IOException;
+
+/**
+ * Thrown by SecretHandler when a secret could not be understood or fetched.
+ */
+public class SecretHandlerException extends RuntimeException {
```
Comment (agavrilov76) [inline flags/src/main/java/org/secondbase/secrets/SecretHandlerException.java:8]: +1

**Review by agavrilov76 — COMMENTED** [review]
Very nice. Only few minor comments

File: flags/pom.xml (hunk 1 above)
Comment (acidmoose) [reply flags/pom.xml:26]: I'll do the same change for vault once this is ok and merged. Don't want to do double change if it's not ok, and to keep the pull request small (well, relatively small...).

File: flags/src/main/java/org/secondbase/flags/Flags.java (hunk 2 above)
Comment (acidmoose) [reply flags/src/main/java/org/secondbase/flags/Flags.java]: You mean multiple SecretHandlers? It should call fetch(args) on all SecretHandler implementations, and the implementations themselves are responsible for fetching all secrets passed. I should add a dummy fetcher for testing inside of Flags component. And another test for multiple secret flags.

Comment (agavrilov76) [reply flags/src/main/java/org/secondbase/flags/Flags.java]: Should we pass 'ret' instead of 'args' as an argument to make it work?

Comment (acidmoose) [reply flags/src/main/java/org/secondbase/flags/Flags.java]: Sorry. I didn't understand the last comment.

Comment (acidmoose) [reply flags/src/main/java/org/secondbase/flags/Flags.java]: Damn, that was a very stupid mistake. Thanks for spotting that one.

**Review by agavrilov76 — APPROVED** [review]

**Review by agavrilov76 — APPROVED** [review]

**Pull request merged**